softlimit.8
.TH softlimit 8
.SH NAME
softlimit \- runs another program with new resource limits.
.SH SYNOPSIS
.B softlimit
[
.I opts
]
.I child
.SH DESCRIPTION
.I opts
is a series of getopt-style options.
.I child
consists of one or more arguments. 

.B softlimit
sets soft resource limits as specified by
.IR opts .
It then runs
.IR child .
.SH OPTIONS
In each of the following
.IB opts ,
.I n
may be
.BR = ,
indicating that the soft limit should be set equal to the hard
limit, or
.BR -,
indicating that both the soft and hard limits should be reset to unlimited.

.I opts
.B controlling memory use: 

.TP
.B \-m \fIn
Same as
.B \-d
.I n
.B \-s
.I n
.B \-l
.I n
.B \-a
.IR n .
.TP
.B \-d \fIn
Limit the data segment per process to
.I n
bytes.
.TP
.B \-s \fIn
Limit the stack segment per process to
.I n
bytes.
.TP 
.B \-l \fIn
Limit the locked physical pages per process to
.I n
bytes. This option has no effect on some operating systems.
.TP
.B \-a \fIn
Limit the total of all segments per process to
.I n
bytes. This option has no effect on some operating systems. 
.TP
.B \-o \fIn
Limit the number of open file descriptors per process to
.IR n .
This option has no effect on some operating systems. 
.TP
.B \-p \fIn
Limit the number of processes per uid to
.IR n .
.P
.I opts
.B controlling file sizes: 
.TP
.B \-f \fIn
Limit output file sizes to
.I n
bytes.
.TP
.B \-c \fIn
Limit core file sizes to
.I n
bytes.
.P
.B Efficiency
.I opts:
.TP
.B \-r \fIn
Limit the resident set size to
.I n
bytes. This limit is not enforced unless physical memory is full.
.TP
.B \-t \fIn
Limit the CPU time to
.I n
seconds. This limit is not enforced except that the process receives a
SIGXCPU signal after
.I n
seconds. 
.SH ENVIRONMENT
.TP
.B SOFTLIMIT_ALLBYTES
Same as
.B \-a
.TP
.B SOFTLIMIT_COREBYTES
Same as
.B \-c
.TP
.B SOFTLIMIT_DATABYTES
Same as
.B \-d
.TP
.B SOFTLIMIT_FILEBYTES
Same as
.B \-f
.TP
.B SOFTLIMIT_LOCKEDBYTES
Same as
.B \-l
.TP
.B SOFTLIMIT_MEMBYTES
Same as
.B \-m
.TP
.B SOFTLIMIT_OPENFILES
Same as
.B \-o
.TP
.B SOFTLIMIT_PROCS
Same as
.B \-p
.TP
.B SOFTLIMIT_RSSBYTES
Same as
.B \-r
.TP
.B SOFTLIMIT_STACKBYTES
Same as
.B \-s
.TP
.B SOFTLIMIT_CPUSECS
Same as
.B \-t
.SH SEE ALSO
envdir(8),
envini(8),
envuidgid(8),
fghack(8),  
multilog(8),
pgrphack(8),
readproctitle(8),
setlock(8),
setuidgid(8),
setuser(8),
supervise(8),
svc(8),
svok(8),
svscan(8),
svscanboot(8),
svstat(8),
tai64n(8),
tai64nlocal(8)


CHANGES
Added new program setuser to setuid, setenv, plus chdir before exec.

Added new program envini to read environment variables from an INI file.

Added support for adding a prefix in envdir.

Added "once" flag output to svstat.

Fixed sleeper to do a single atomic write.

Added support for resetting limits in softlimit.

Changes in daemontools-encore-1.10
==================================

Added support for fnmatch patterns in multilog.

Increased the multilog file size limit to 2GB.
(Thanks to SATOH Fumiyasu)

Fixed missing separator between system error texts.

Major speed ups in multilog, tai64n, and tai64nlocal.

Added support for signal propagation and disabling setsid behavior in
supervise.  (Thanks to Alan Grow)

Changes in daemontools-encore-1.06
==================================

Added options to svstat to control checking the log script service.

Added -l option to svc to explicitly control the main service.

Fixed race issuing svc commands to log and then main service.

Changed the "safely written" option for multilog to "w".

Fixed daemontools leaking the status file descriptor to its children.

Fixed svc sending commands to both the log and main service.

Changes in daemontools-encore-1.05
==================================

Added options to svup to control checking the log script service.

Eliminated some compiler warnings.

Changes in daemontools-encore-1.04
==================================

Fixed bug in svscan preventing it from starting services with a "log"
script instead of directory.

Changes in daemontools-encore-1.03
==================================

Added -s option to setuidgid to set the supplemental GIDs.
(Thanks to SATOH Fumiyasu)

Changes in daemontools-encore-1.02
==================================

If a supervise is managing both a main and log service, issuing
"svc -dx" on the service will cause the log to stop after the main
service exits.

Changes between daemontools-0.76 and daemontools-encore
=======================================================

Added support to supervise/svc for sending SIGQUIT/USR1/USR2/WINCH.
(original patch by Wayne Marshall, based on patch by Scott Gifford)

Modified supervise to automatically put the forked service into its own
process group to facilitate sending signals to the group.

Added support to supervise/svc for sending signals to the group.

Added support to softlimit for reading limits from environment variables.

Modified supervise to run "start" once before executing "run" to allow
for single-shot initialization, and to run "stop" after "run" exits for
the last time.  The presence of "run" is also optional.

Modified supervise to stop trying to restart the service if it exits 100.
(semantics borrowed from qmail control files)

Added support to multilog for "accustamp" time stamps.
(derived from patch by Matthias Andree)

Added support to supervise for storing the status files in an alternate
directory specified by $SUPERVISEDIR.

Modified multilog to support specifying an alternate "safely written"
filename extension (useful for compressed logs).

Added extended state information to the supervise/status file and svstat.

Added support to supervise for a "notify" script, executed whenever a
subprocess is started or exits.  If present, it is passed the script
name, "start" or "exit", the PID, and the exit code.

Added support for piping to "log" within supervise.  This requires "log"
to be an executable file (instead of a directory as for svscan).

Modified supervise to overwrite status instead of creating a new file.
This allows the "supervise" subdirectory to be non-writeable.

Added a "svup" tool to determine if a service is running.

Modified svscan to optionally redirect all output through a service.
(adapted from http://www.gluelogic.com/code/daemontools/daemontools-0.76-gps-logging-0.01.patch)
Note that this feature works differently from the GlueLogic
modification.  See the svscan(8) man page for details.


softlimit.c
#include <sys/types.h>
#include <time.h>
#include <sys/time.h>
#include <sys/resource.h>
#include "env.h"
#include "pathexec.h"
#include "sgetopt.h"
#include "strerr.h"
#include "scan.h"
#include "str.h"

#define FATAL "softlimit: fatal: "

static void die_usage(void)
{
  strerr_die1x(100,"softlimit: usage: softlimit [-a allbytes] [-c corebytes] [-d databytes] [-f filebytes] [-l lockbytes] [-m membytes] [-o openfiles] [-p processes] [-r residentbytes] [-s stackbytes] [-t cpusecs] child");
}

static void doit(int resource,const char *arg)
{
  unsigned long u;
  struct rlimit r;

  if (getrlimit(resource,&r) == -1)
    strerr_die2sys(111,FATAL,"getrlimit failed");

  if (str_equal(arg,"="))
    r.rlim_cur = r.rlim_max;
  else if (str_equal(arg,"-"))
    r.rlim_cur = r.rlim_max = RLIM_INFINITY;
  else {
    if (arg[scan_ulong(arg,&u)]) die_usage();
    r.rlim_cur = u;
    if (r.rlim_cur > r.rlim_max)
      r.rlim_cur = r.rlim_max;
  }

  if (setrlimit(resource,&r) == -1)
    strerr_die2sys(111,FATAL,"setrlimit failed");
}

static void doenv(int resource,const char *name)
{
  const char *arg;
  if ((arg = env_get(name)) != 0)
    doit(resource,arg);
}

static void doallenv(void)
{
#ifdef RLIMIT_AS
  doenv(RLIMIT_AS,"SOFTLIMIT_ALLBYTES");
  doenv(RLIMIT_AS,"SOFTLIMIT_MEMBYTES");
#endif
#ifdef RLIMIT_VMEM
  doenv(RLIMIT_VMEM,"SOFTLIMIT_ALLBYTES");
  doenv(RLIMIT_VMEM,"SOFTLIMIT_MEMBYTES");
#endif
#ifdef RLIMIT_CORE
  doenv(RLIMIT_CORE,"SOFTLIMIT_COREBYTES");
#endif
#ifdef RLIMIT_DATA
  doenv(RLIMIT_DATA,"SOFTLIMIT_DATABYTES");
  doenv(RLIMIT_DATA,"SOFTLIMIT_MEMBYTES");
#endif
#ifdef RLIMIT_FSIZE
  doenv(RLIMIT_FSIZE,"SOFTLIMIT_FILEBYTES");
#endif
#ifdef RLIMIT_MEMLOCK
  doenv(RLIMIT_MEMLOCK,"SOFTLIMIT_LOCKEDBYTES");
  doenv(RLIMIT_MEMLOCK,"SOFTLIMIT_MEMBYTES");
#endif
#ifdef RLIMIT_STACK
  doenv(RLIMIT_STACK,"SOFTLIMIT_STACKBYTES");
  doenv(RLIMIT_STACK,"SOFTLIMIT_MEMBYTES");
#endif
#ifdef RLIMIT_NOFILE
  doenv(RLIMIT_NOFILE,"SOFTLIMIT_OPENFILES");
#endif
#ifdef RLIMIT_OFILE
  doenv(RLIMIT_OFILE,"SOFTLIMIT_OPENFILES");
#endif
#ifdef RLIMIT_NPROC
  doenv(RLIMIT_NPROC,"SOFTLIMIT_PROCS");
#endif
#ifdef RLIMIT_RSS
  doenv(RLIMIT_RSS,"SOFTLIMIT_RSSBYTES");
#endif
#ifdef RLIMIT_CPU
  doenv(RLIMIT_CPU,"SOFTLIMIT_CPUSECS");
#endif
}

int main(int argc,const char *const *argv,const char *const *envp)
{
  int opt;

  doallenv();
  while ((opt = getopt(argc,argv,"a:c:d:f:l:m:o:p:r:s:t:")) != opteof)
    switch(opt) {
      case '?':
	die_usage();
      case 'a':
#ifdef RLIMIT_AS
        doit(RLIMIT_AS,optarg);
#endif
#ifdef RLIMIT_VMEM
        doit(RLIMIT_VMEM,optarg);
#endif
        break;
      case 'c':
#ifdef RLIMIT_CORE
        doit(RLIMIT_CORE,optarg);
#endif
        break;
      case 'd':
#ifdef RLIMIT_DATA
        doit(RLIMIT_DATA,optarg);
#endif
        break;
      case 'f':
#ifdef RLIMIT_FSIZE
        doit(RLIMIT_FSIZE,optarg);
#endif
        break;
      case 'l':
#ifdef RLIMIT_MEMLOCK
        doit(RLIMIT_MEMLOCK,optarg);
#endif
        break;
      case 'm':
#ifdef RLIMIT_DATA
        doit(RLIMIT_DATA,optarg);
#endif
#ifdef RLIMIT_STACK
        doit(RLIMIT_STACK,optarg);
#endif
#ifdef RLIMIT_MEMLOCK
        doit(RLIMIT_MEMLOCK,optarg);
#endif
#ifdef RLIMIT_VMEM
        doit(RLIMIT_VMEM,optarg);
#endif
#ifdef RLIMIT_AS
        doit(RLIMIT_AS,optarg);
#endif
	break;
      case 'o':
#ifdef RLIMIT_NOFILE
        doit(RLIMIT_NOFILE,optarg);
#endif
#ifdef RLIMIT_OFILE
        doit(RLIMIT_OFILE,optarg);
#endif
        break;
      case 'p':
#ifdef RLIMIT_NPROC
        doit(RLIMIT_NPROC,optarg);
#endif
        break;
      case 'r':
#ifdef RLIMIT_RSS
        doit(RLIMIT_RSS,optarg);
#endif
        break;
      case 's':
#ifdef RLIMIT_STACK
        doit(RLIMIT_STACK,optarg);
#endif
        break;
      case 't':
#ifdef RLIMIT_CPU
        doit(RLIMIT_CPU,optarg);
#endif
        break;
    }

  argv += optind;
  if (!*argv) die_usage();

  pathexec_run(*argv,argv,envp);
  strerr_die3sys(111,FATAL,"unable to run ",*argv);
}
